cpp/src/arrow/array/data.cc

```diff
@@ -144,6 +144,8 @@ std::shared_ptr<ArrayData> ArrayData::Slice(int64_t off, int64_t len) const {
   } else {
     copy->null_count = null_count != 0 ? kUnknownNullCount : 0;
   }
+  for (auto& child : copy->child_data) 
+    child = child->Slice(copy->offset, copy->length);
```

Comment:
I think this breaks the contract of a sliced array. Please check the docstring of `ArrayData`
```
  // The logical start point into the physical buffers (in values, not bytes).
  // Note that, for child data, this must be *added* to the child data's own offset.
  int64_t offset = 0;
```

Reply:
cc @lidavidm @westonpace

Reply:
Yes, the expectation *in the C++ implementation* is that slicing an array is a metadata-only operation that just adjusts the "offset" of an array, and does not affect child arrays. That means code directly manipulating child arrays must (recursively) account for parent array offsets.

However, it may be good to document how to 'materialize' or slice the child array(s) to account for this. I'm not aware of a convenient way to do this.

Reply:
Can we provide arraydata with slice operations on child_data?

Reply:
> Can we provide arraydata with slice operations on child_data?

What is the use case? Both `ArrayData` and `Buffer` do not provide slice operations so it would be non-trivial to do this.

Reply:
> > Can we provide arraydata with slice operations on child_data?
> 
> What is the use case? Both `ArrayData` and `Buffer` do not provide slice operations so it would be non-trivial to do this.

case: I have an two phase avg.

phase1: returns an intermediate result struct array<sum, count>.

phase2: input: struct array<sum, count>, output: float8, in Consume, I imitate SumImpl to implement addition to struct<sum, count>，split it to sum_arr and count_arr.

just like SumImpl,  `this->sum += SumArray<CType, SumCType, SimdLevel>(*data);` data is from :`const auto& data = batch[0].array();`. it is an arraydata not array.

so for our phase 2 consum, input data is a struct array data, code like this:
```cpp
auto arraydata = *(batch[0].array());
ConsumeSumArray(arraydata.child_data[0]);
ConsumeCountArray(arraydata.child_data[1]);
```

But in the grouping scenario, I may have my batch sliced, so arraydata should also be sliced, so I have to do this here

```
auto array_data = batch[0].array();
auto struct_array = std::make_shared<StructArray>(array_data);
ConsumeSumArray(struct_array->field(0)->data());
ConsumeCountArray(struct_array->field(1)->data());
```

Reply:
The essence is that the agg function operates on arraydata. If I need a slice, I must call it explicitly or construct an array before using it. Is it reasonable?

Reply:
In this case, you should explicitly pass around the parent offset or adjust the child offsets. 

It may be reasonable to provide a method on Array that can pre-adjust the children for you (without actually copying the data).

Reply:
OK, thank you

Reply:
> 

good idea1